### PR TITLE
[C-843] Native analytics sagas

### PR DESCRIPTION
**Pull request by sliptype**
### Description

> May be easiest to review by commit bc of all the path changes

* Move analytics actions and sagas into web common
* Add `analytics` object into storeContext
* Add `webAnalyticsSagas` that handle page view tracking (still emitting messages to mobile for now)
* Add common analytics sagas to mobile (aren't being used yet)
* Move mobile `utils/analytics` to `services/analytics`

### Dragons

Had to make the `analytics` interface consistent between web and mobile to use in the storeContext. Decided to conform to 

```ts
{
  eventName: string
  properties?: Record<string, any>
}
```

because mobile was already using that format everywhere and calling `track` directly, whereas web was mostly using the `useRecord` hook so the change only needed to happen in one place. I added an issue to make the readdress the analytics implementation and make it consistent across the board: https://linear.app/audius/issue/C-848/make-analytics-implementation-consistent-across-web-and-mobile

### How Has This Been Tested?

Tested analytics on web and mobile including page views and payloads

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

### CODE REVIEW

File: packages/mobile/src/store/sagas.ts [hunk 1]
```diff
@@ -12,6 +13,7 @@ export default function* rootSaga() {
   const sagas = [
     // config
     ...backendSagas(),
+    ...analyticsSagas(),
```
Comment (dylanjeffers) [inline packages/mobile/src/store/sagas.ts:16]: cool! so we want to have these around for now then?

File: packages/web/src/common/services/audius-backend/AudiusBackend.ts [hunk 2]
```diff
@@ -234,8 +234,7 @@ type AudiusBackendParams = {
   onLibsInit: (libs: any) => void
   recaptchaSiteKey: Maybe<string>
   recordAnalytics: (
-    event: string,
-    properties?: Record<string, any>,
+    event: { eventName: string; properties?: Record<string, any> },
```
Comment (dylanjeffers) [inline packages/web/src/common/services/audius-backend/AudiusBackend.ts:237]: do we want to use `Record<string, unknown>` i think that is technically the best version of "any object"

File: packages/web/src/common/services/audius-backend/AudiusBackend.ts [hunk 3]
```diff
@@ -541,16 +546,22 @@ export const audiusBackend = ({
     secondaries: string[],
     reason: string
   ) {
-    recordAnalytics(Name.CREATOR_NODE_SELECTION, {
-      endpoint: primary,
-      selectedAs: 'primary',
-      reason
+    recordAnalytics({
+      eventName: Name.CREATOR_NODE_SELECTION,
```
Comment (dylanjeffers) [inline packages/web/src/common/services/audius-backend/AudiusBackend.ts:550]: wow epic job changing all these

File: packages/web/src/common/store/analytics/sagas.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+} from 'common/store/analytics/actions'
+
+function* trackEventAsync(action: TrackEvent) {
+  const analytics = yield* getContext('analytics')
+  const { callback, eventName, ...properties } = action
+  yield call(
+    analytics.track,
+    {
+      eventName,
+      properties: omit(properties, 'type')
```
Comment (dylanjeffers) [inline packages/web/src/common/store/analytics/sagas.ts:19]: awesome stuff, so do we just want to omit type in the destructuring?

File: packages/web/src/common/store/storeContext.ts [hunk 5]
```diff
@@ -10,6 +14,16 @@ export type CommonStoreContext = {
   getLocalStorageItem: (key: string) => Promise<string | null>
   setLocalStorageItem: (key: string, value: string) => Promise<void>
   getFeatureEnabled: (flag: FeatureFlags) => Promise<boolean | null>
+  analytics: {
+    init: () => Promise<void>
+    track: (event: AnalyticsEvent, callback?: () => void) => Promise<void>
+    identify: (
+      handle: string,
+      traits?: Record<string, any>,
```
Comment (dylanjeffers) [inline packages/web/src/common/store/storeContext.ts:22]: same here just regarding `<string, any>` vs `<string, unknown>`

File: packages/web/src/components/edit-folder-modal/EditFolderModal.tsx [hunk 6]
```diff
@@ -12,13 +12,13 @@ import { useDispatch } from 'react-redux'
 
 import { useModalState } from 'common/hooks/useModalState'
 import { getPlaylistLibrary } from 'common/store/account/selectors'
+import { make, useRecord } from 'common/store/analytics/actions'
```
Comment (dylanjeffers) [inline packages/web/src/components/edit-folder-modal/EditFolderModal.tsx:15]: does it makes sense to have `useRecord` in actions? i guess it's a special action? and looks like that's what was there before anyway so nvm

File: packages/web/src/components/notification/store/sagas.ts [hunk 7]
```diff
@@ -671,10 +670,6 @@ export function* markAllNotificationsViewed() {
   const audiusBackendInstance = yield* getContext('audiusBackendInstance')
   yield* call(waitForBackendSetup)
   yield* call(audiusBackendInstance.markAllNotificationAsViewed)
-  if (NATIVE_MOBILE) {
```
Comment (dylanjeffers) [inline packages/web/src/components/notification/store/sagas.ts:674]: i forgot, we still want this yeah?

File: packages/web/src/pages/audio-rewards-page/components/modals/CognitoModal.tsx [hunk 8]
```diff
@@ -62,17 +63,21 @@ export const CognitoModal = () => {
...
+          )
           break
         case 'closed':
           dispatch(setCognitoFlowStatus({ status: CognitoFlowStatus.CLOSED }))
-          track(Name.REWARDS_CLAIM_FINISH_COGNITO_FLOW, {
-            handle,
-            source
-          })
+          record(
+            make(Name.REWARDS_CLAIM_FINISH_COGNITO_FLOW, {
```
Comment (dylanjeffers) [inline packages/web/src/pages/audio-rewards-page/components/modals/CognitoModal.tsx:76]: just wondering if make should just be a part of record at some point?

**Review by dylanjeffers — APPROVED** [review]
all makes good sense, nice job

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/sk-c-843-native-analytics-sagas

File: packages/mobile/src/store/sagas.ts (hunk 1 above)
Comment (sliptype) [reply packages/mobile/src/store/sagas.ts:16]: Yeah, they aren't touching any state. Don't do anything other than call amplitude essentially so it's fine to have them

File: packages/web/src/common/services/audius-backend/AudiusBackend.ts (hunk 2 above)
Comment (sliptype) [reply packages/web/src/common/services/audius-backend/AudiusBackend.ts:237]: Yup, can actually use `AnalyticsEvent` here

File: packages/web/src/common/services/audius-backend/AudiusBackend.ts (hunk 3 above)
Comment (sliptype) [reply packages/web/src/common/services/audius-backend/AudiusBackend.ts:550]: Thanks! idk if this is the final interface we should use but updating web was the quickest route

File: packages/web/src/components/edit-folder-modal/EditFolderModal.tsx (hunk 6 above)
Comment (sliptype) [reply packages/web/src/components/edit-folder-modal/EditFolderModal.tsx:15]: No it doesn't make sense, we can address in the other analytics issue I created

File: packages/web/src/pages/audio-rewards-page/components/modals/CognitoModal.tsx (hunk 8 above)
Comment (sliptype) [reply packages/web/src/pages/audio-rewards-page/components/modals/CognitoModal.tsx:76]: Definitely think so, can address as part of the other issue

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/sk-c-843-native-analytics-sagas

**Pull request merged**